frontend/packages/dev-console/src/components/topology2/components/nodes/KnativeService.tsx

```diff
@@ -36,10 +39,34 @@ const KnativeService: React.FC<EventSourceProps> = ({
   dragging,
   regrouping,
 }) => {
-  const trafficAnchor = useSvgAnchor(AnchorEnd.source, 'revision-traffic');
+  const { data } = element.getData();
+  const hasDataUrl = !!data.url;
+  useAnchor(
+    React.useCallback(
+      (node: Node) => {
+        class TrafficAnchor extends AbstractAnchor {
+          getLocation(reference: Point): Point {
+            const bb = this.getOwner().getBounds();
+            // center point is top right corner
+            const center = new Point(bb.right(), bb.y);
+            if (hasDataUrl) {
+              const size = DECORATOR_RADIUS * 2;
+              return getEllipseAnchorPoint(center, size, size, reference);
+            }
+
+            // if no data url, then there is no decorator edge to anchor to
+            return center;
+          }
+        }
```

Comment:
Is `TrafficAnchor` worthy of its own file rather than defined inline?

Reply:
Create a separate file.